tests/testsuite/build_script.rs

```diff
@@ -2207,6 +2215,206 @@ fn test_duplicate_deps() {
     p.cargo("build").run();
 }
 
+#[cargo_test]
+fn test_duplicate_shared_deps_native() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [project]
+                name = "foo"
+                version = "0.1.0"
+                authors = []
+                build = "build.rs"
+
+                [dependencies]
+                bar = { path = 'bar' }
+
+                [build-dependencies]
+                bar = { path = 'bar' }
+            "#,
+        )
+        .file(
+            "src/main.rs",
+            r#"
+                extern crate bar;
+                fn main() { bar::do_nothing() }
+            "#,
+        )
+        .file(
+            "build.rs",
+            r#"
+                extern crate bar;
+                use std::env;
+                fn main() {
+                    bar::do_nothing();
+                    assert_eq!(env::var("DEP_FOO_FOO").unwrap(), "bar");
+                    assert_eq!(env::var("CARGO_BUILD_TYPE").unwrap(), "native");
+                }
+            "#,
+        )
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [project]
+                name = "bar"
+                version = "0.1.0"
+                authors = []
+                links = "foo"
+                build = "build.rs"
+            "#,
+        )
+        .file(
+            "bar/build.rs",
+            r#"
+                use std::env;
+                fn main() {
+                    println!("cargo:foo=bar");
+                    assert_eq!(env::var("CARGO_BUILD_TYPE").unwrap(), "native");
```

Comment:
This is the bug that I was mentioning. This build script is run once instead of twice, although there are two possible compilations in this scenario. 

We want to move towards a world where `cargo build` is the same as `cargo build --target $host`, but this is a case where they are not the same.

Reply:
> This is the bug that I was mentioning. This build script is run once instead of twice, although there are two possible compilations in this scenario.

But at least for this `CARGO_BUILD_TYPE` it's still giving expected results...so not sure why that bug would be an issue for the `CARGO_BUILD_TYPE` feature.

> We want to move towards a world where `cargo build` is the same as `cargo build --target $host`, but this is a case where they are not the same.

Isn't this still somewhat problematic for cases where one needs different logic for cross compilations due to the inability to execute binaries built for the target on the host?

Reply:
For `CARGO_BUILD_TYPE` this would be an incorrect value. The model that Cargo wants to get to is that every single compilation is a "cross" compilation where there's a clear distinction between host/target artifacts and you can separately configure both. That means that the build script here should either run once in "cross" mode or twice, once in "cross" and once in "native"

(FWIW I'm not personally a fan of the "build type" and "cross/native" terminology to get exposed here. I think their purpose is fine but eventually I'd want to bikeshed on changing the interface. For now I'd like to get agreement on what needs to be implemented first though)

If `--target` doesn't equal the host platform, then the binary naturally won't run on the host, but if it's omitted or equals the `$host` then it's up to build configuration as to whether the binary runs on the host. By default the output of `cargo build` would be able to run on the host.

Reply:
> The model that Cargo wants to get to is that every single compilation is a "cross" compilation where there's a clear distinction between host/target artifacts and you can separately configure both.

While being able to separately configure host/targets is needed for cross compilation it's not really exclusive to cross compilation scenarios and can be also done for native compilation(for example using different optimization flags for build scripts vs the actual target binaries).

> That means that the build script here should either run once in "cross" mode or twice, once in "cross" and once in "native"

Best I can tell cross vs native is a somewhat different issue from host vs target configuration. In rust/cargo they seem to be fairly closely tied but for build scripts that need to say build against non-rust dependencies like python there's additional implications where having this env variable is more important so that the configuration logic works correctly.

> (FWIW I'm not personally a fan of the "build type" and "cross/native" terminology to get exposed here. I think their purpose is fine but eventually I'd want to bikeshed on changing the interface. For now I'd like to get agreement on what needs to be implemented first though)

Yeah, I figured that would be the case, this was just the first thing I came up with.

> If `--target` doesn't equal the host platform, then the binary naturally won't run on the host, but if it's omitted or equals the `$host` then it's up to build configuration as to whether the binary runs on the host.

Some build systems like meson have [wrapper capabilities](https://mesonbuild.com/Cross-compilation.html#binaries) to work around these limitations. There's a good bit of unusual configurations you can get here in general, for example in buildroot we have to use a qemu wrapper for gobject-introspection cross compilation(although that actually isn't implemented using the native meson exe wrapper but rather by overriding specific gobject-introspection executables with buildroot managed wrapper scripts, it's rather complex). But I think that's somewhat separate from this feature which is mostly just for indicating if the target was overridden to the build script, what the build script actually does with that info would be highly specific to the build script use case.

> By default the output of `cargo build` would be able to run on the host.

Sure, and in the pyo3 case I think exposing if it was run with default vs an explicit target to the build script is helpful, however it does seem there may be some unclear edge cases here so not entirely sure if this is the best approach. Right now pyo3 just uses an `target == host` check(which is obviously broken when cross compiling for the same target arch as the host) to determine if it is cross compiling, although there are some override env variables that I added there for now which can override that check. I figured at least this way the build script can know that the target was overridden.

Reply:
I feel like we're roughly talking about the same thing but it seems to be straying from the original comment I made about the value of `native` on this line I don't think is correct. To restate some properties I believe Cargo wants to achieve, but doesn't today:

* `cargo build` should be the same as `cargo build --target $host`
* Cargo should support separate host/target configuration.
  * As corrolary of the above point `cargo build` should use the host/target configuration as it does in the same manner of `cargo build --target $host`
* All compilations should be "cross" compilations (see the first point) but the term "cross" isn't quite right because `cargo build` isn't building anything for a different architecture.

Cargo does not today treat `cargo build` the same as `cargo build --target $host`. Cargo has recently picked up support for separate host/target configuration but it's going to be blocked on fixing this.

Projects like pyo3 which want to link to different python interpreters or something like that will always need some means of auxiliary configuration. If Cargo informs pyo3 that the build is happening for the host ("native" in this PR more-or-less) then it probably wants to disregard the configuration and look at the host to see what it can do. If Cargo informs pyo3 that the build is happening for the target (roughly "cross" as-is here, but not completely because this should show up in the `cargo build` case too) then pyo3 would look for configuration. If it finds no configuration and host==target then it would assume to link to the current system.

Reply:
> I feel like we're roughly talking about the same thing but it seems to be straying from the original comment I made about the value of `native` on this line I don't think is correct.

Well I'm seeing essentially two different things here, making cargo's internal logic handle native builds in the same way as cross is the first one, but that's not really what I'm trying to do here. The issue I'm trying to address here is regarding passing of invocation flag information on whether a specific target was specified when invoking cargo to the build scripts as they may integrate with externally managed dependencies(like python in the pyo3 case) that need to be informed of that.

> Cargo does not today treat `cargo build` the same as `cargo build --target $host`. Cargo has recently picked up support for separate host/target configuration but it's going to be blocked on fixing this.

I don't think this is really blocked by support for splitting host/target configurations at all as this should not change based on that, essentially all I'm trying to do here is inform the build script if the user set a custom target flag for the cargo build, whether the target config should apply to the host build script is a separate concern and should be independent of whether or not a target flag was passed to cargo, if the build script doesn't actually use the `cross` vs `native` information for anything then there would not be any behavior change. We may in addition want to add a variable that informs the build script if a target config applies to the host so that the build script heuristics can be further improved, and that would actually be dependent on the separate host/target configuration changes.

> Projects like pyo3 which want to link to different python interpreters or something like that will always need some means of auxiliary configuration.

Yeah, so the idea is that this is just additional information to improve the build script auxiliary configuration heuristics. Keep in mind for the pyo3 case the target configuration params are effectively chained from a setuptools plugin, we mostly just want to preserve that information when it gets passed to the build script.

Essentially a typical cross compile invoked from the meta build system like buildroot would look something like this I think:

1. buildroot invokes host-python build against a python package like python-cryptography with cross config in env or cargo config file with config file location in env
2. host-python invokes setuptools-rust
3. setuptools-rust running in host-python interpreter invokes cargo while forwarding the target config to cargo
4. cargo invokes the pyo3 build script while passing target configs to build script
5. pyo3 build script picks up target configuration passed by cargo during execution
6. pyo3 build script reads python sysconfigdata to determine how to link against target-python based on information from cargo(this particular change is helpful as it helps the build script know how to appropriately pick up target-python's sysconfigdata)
7. the extension module is compiled by cargo configured against target-python's sysconfigdata params

What I'm trying to ensure here is that target flag invocation information is available through the entire cross compilation call chain here. Keep in mind that we don't invoke cargo directly when using pyo3 typically, cargo is effectively always being managed by setuptools-rust which in turn may or may not be managed by another meta build system(buildroot in this case).

In addition it's conceivable that the cargo build script may in addition need to do runtime python invocations for configuration and in that case it would need to properly differentiate the host-python from the target-python interpreter depending on build requirements, we even have some exceptionally complex/exotic scenarios in buildroot involving GIR code-generation which requires target python invocations in qemu wrappers that may require additional special cased handling. So a build script may need to be using both host and target configs depending on the requirements.

> If Cargo informs pyo3 that the build is happening for the host ("native" in this PR more-or-less) then it probably wants to disregard the configuration and look at the host to see what it can do.

Well this seems to be highly dependent on the specific cross compilation requirements of the build script I think, which is why I didn't want it to directly change any of the other cross or host config variables that get passed to the build script or any behavior directly.

> If Cargo informs pyo3 that the build is happening for the target (roughly "cross" as-is here, but not completely because this should show up in the cargo build case too) then pyo3 would look for configuration.

Well the build script looks like it really should be able to know how the cargo build was invoked, even if cargo's own internal logic should not change based on if a target flag was set. Essentially the `cross` param is a hint to the build script that it may in some cases depending on requirements would need to adjust its configuration logic due to a target flag having been passed to cargo. This hint could also be used by the build script merely for doing sanity checks that emit warnings.

This isn't really intended to be used for host vs target `TargetConfig` build target variant differentiation but rather to inform the build script how cargo was originally invoked.

> If it finds no configuration and host==target then it would assume to link to the current system.

Yeah, so one of the motivations here is to provide a way for the build script to better deal with this particular edge case, currently pyo3 is using this sort of `host==target` comparison, however this is unreliable since `host==target` alone is not enough to determine if a target flag was passed to cargo or if one is cross compiling in cases where the host and target use the same triple, if a target flag was passed to cargo it's highly likely that the user is cross compiling and the build script should thus be informed of that so it can appropriately handle the situation or at least warn the user that additional information may be required for a successful build, it would generally be expected that the build script provide escape hatch env variables regardless for overriding behavior.

Hopefully that makes sense, I'm trying to cover additional background info for clarity as some of the more exotic scenarios may be non-obvious.